src/App/Pages/Accounts/LoginSsoPageViewModel.cs

```diff
@@ -0,0 +1,177 @@
+            }
+            var rememberOrgIdentifier = await _storageService.GetAsync<bool?>(Keys_RememberOrgIdentifier);
+            RememberOrgIdentifier = rememberOrgIdentifier.GetValueOrDefault(true);
+        }
+
+        public async Task LogInAsync()
+        {
+            if (Connectivity.NetworkAccess == NetworkAccess.None)
+            {
+                await _platformUtilsService.ShowDialogAsync(AppResources.InternetConnectionRequiredMessage,
+                    AppResources.InternetConnectionRequiredTitle);
+                return;
+            }
+            if (string.IsNullOrWhiteSpace(OrgIdentifier))
+            {
+                await _platformUtilsService.ShowDialogAsync(
+                    string.Format(AppResources.ValidationFieldRequired, AppResources.OrgIdentifier),
+                    AppResources.AnErrorHasOccurred,
+                    AppResources.Ok);
+                return;
+            }
+
+            await _deviceActionService.ShowLoadingAsync(AppResources.LoggingIn);
+
+            var passwordOptions = new PasswordGenerationOptions(true);
+            passwordOptions.Length = 64;
+
+            var codeVerifier = await _passwordGenerationService.GeneratePasswordAsync(passwordOptions);
+            var codeVerifierHash = await _cryptoFunctionService.HashAsync(codeVerifier, CryptoHashAlgorithm.Sha256);
+            var codeChallenge = CoreHelpers.Base64UrlEncode(codeVerifierHash);
+            await _storageService.SaveAsync(Constants.SsoCodeVerifierKey, codeVerifier);
+
+            var state = await _passwordGenerationService.GeneratePasswordAsync(passwordOptions);
+            await _storageService.SaveAsync(Constants.SsoStateKey, state);
+
+            var redirectUri = "bitwarden://sso-callback";
+
+            var url = _apiService.IdentityBaseUrl + "/connect/authorize?" +
+                      "client_id=" + _platformUtilsService.IdentityClientId + "&" +
+                      "redirect_uri=" + Uri.EscapeDataString(redirectUri) + "&" +
+                      "response_type=code&scope=api%20offline_access&" +
+                      "state=" + state + "&code_challenge=" + codeChallenge + "&" +
+                      "code_challenge_method=S256&response_mode=query&" +
+                      "domain_hint=" + Uri.EscapeDataString(OrgIdentifier);
+
+            WebAuthenticatorResult authResult = null;
+            bool cancelled = false;
+            try
+            {
+                authResult = await WebAuthenticator.AuthenticateAsync(new Uri(url),
+                    new Uri(redirectUri));
+            }
+            catch (TaskCanceledException e)
+            {
+                await _deviceActionService.HideLoadingAsync();
+                cancelled = true;
+            }
+            if (!cancelled)
+            {
+                if (authResult != null && authResult.Properties.TryGetValue("code", out var code))
```

Comment:
According to spec, you are also suppose to verify that the state returned is the same as what you originally passed.